### PR TITLE
Cote event generic

### PR DESCRIPTION
**Pull request by roblabat**
Please fill in this template.

  Use a meaningful title for the pull request. Include the name of the package modified.
  Test the change in your own code. (Compile and run.)
  Add or edit tests to reflect the change. (Run with npm test.)
  Follow the advice from the readme.
  Avoid common mistakes.
  Run npm run lint package-name (or tsc if no tslint.json is present).
If changing an existing definition:

  Provide a URL to documentation or source code which provides context for the suggested changes: <>
  Increase the version number in the header if appropriate.
  If you are making substantial changes, consider adding a tslint.json containing { "extends": "dtslint/dt.json" }.
This PR adds a GenericEvent type and make it default in generic type functions like responder.on.
This Generic type allow the use of custom event properties without having to define an interface and so simplify the transition from js to ts in a project using [cote](https://github.com/dashersw/cote) without changing the behavior for people using typings on those generically typed functions.

As raised by @makepost in PR [29159](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29159) a simple 
`interface Event {
  type: string;
  [key: string]: any;
}`

will not suits our needs as it would stop raising error on typed usage of those functions. But I think keeping the old behavior that raise error even on an untyped usage on those function is not the good way as it make the migration between ts and js harder by forcing developpers to update all cote projects with custom interfaces. 
Maybe this solution could be a good compromise.

### CODE REVIEW

**Comment by makepost** [comment]
`<T extends Event = GenericEvent>` seems equivalent to `<T extends Event = any>`, which isn't okay in this context. Having developers update their services with custom interfaces is the point of using Cote with TypeScript; not doing that loses the benefits of TypeScript. It's good that compiler tells you all places where you have a proper type missing, and you can temporarily mass-replace `event,` with `event: any /* TODO: Interface */,` in your app code if you really can't add types at the moment but want to enable TypeScript nonetheless.

Thanks for making a change to tslint.json, bringing it to attention that @andy-ms has implemented the `no-unnecessary-generics` rule and added it DefinitelyTyped-wide, but it didn't become apparent earlier because disabled in existing typings in #19564. In fact you'd be better off increasing the strictness by enabling this rule in typings for Cote, dropping the explicit generics where I've used them only once, and simplifying the type from `<T extends Event>(event: T,` to `(event: Event,` as the [rule documentation](https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md) suggests. What do you think?

@dulichan Hi, please have a look at this too as a long-time TypeScript user of Cote.

**Comment by roblabat** [comment]
`<T extends Event = GenericEvent>` is not strictly equivalent to `<T extends Event = any>` as it add the field 'type' but the idea behind it is quite the same effectively (the difference is much more visible on intellisense implementations but it's the same for errors).

I really have a problem making typings that assumes more on data that what we can sure of. In our case the package doesn't return a `{type: string}` argument it return a `{ type: string, [key: string]: any }`. Effectively mass replacing `event` by `event: any /* TODO: Class */` could be a temporary solution for migration (assuming that all your variables in this position is called 'event' and you nether use the world 'event' somewhere else, which is nether true so you will nether be able to do that automatically and will be force to do it manually which can be a really painful job...).
But I think the worst is that you assume you don't have any other properties even if it's not true, most of the time it's not a problem as your code will not try to use it but it can be a big problem if for some reason you use something like a foreach function. Effectively in that case the for each function will be called on properties that you don't want it to be called and typescript will tell you that those properties doesn't exist even if it's false, you could then have an injection leak hidden by typescript. (do not forget that typescript typings or use only before compilation and nether on run time so all objects are in fact any during run time, considering that typings should only be use on variables that we can trust and nether on an external input which should be assume as any).

I think the best way of using cote with typescript (and it's quite the same in js) is to validate incoming messages against a schema everytime and throwing errors if they don't match what you expect, you can so type them after your schema validation but not before it so incoming arguments must be an any (or at least you can assume the type property as cote won't be able to route the message without it and will already throw an error if you don't have this field) in the signature.

Effectively that's a good point to add `no-unnecessary-generics`, I will add that to this PR.

**Comment by roblabat** [comment]
That's what I do, I'm not discussing if it's possible to do that on not. Just if it's the best approach or not.
I think you should be free to do make incoming variables typed or not without typing everything as any which is quite an anti-pattern. Then if you are confident enough in incoming data you can type it and assume it is the type you want but if you are not confident you aren't force to type it.

I also don't see why it could be a problem to make GenericEvent the defaut type as if you do so only code with untyped data will be affected. But currently using the Event type without typing it is useless as it only have the type property that can be inferred by the simple fact that you have been routed the function. in conclusion adding this generic type only allows us to use an other approach of typings without changing anything for people using the previous one.

**Comment by makepost** [comment]
It changes exactly what you're saying, untyped becoms the allowed default, and that's what I'm objecting to. The primary goal of the Event type is to make untyped data an error and help you get rid of it. Untyped should be opt-in per case. Let's see if somebody else differs on this position.

**Comment by andy-ms** [comment]
> But I think the worst is that you assume you don't have any other properties even if it's not true, most of the time it's not a problem as your code will not try to use it but it can be a big problem if for some reason you use something like a foreach function. Effectively in that case the for each function will be called on properties that you don't want it to be called and typescript will tell you that those properties doesn't exist even if it's false, you could then have an injection leak hidden by typescript

Send your :+1: to Microsoft/TypeScript#12936

File: types/cote/index.d.ts [hunk 1]
```diff
@@ -171,7 +174,7 @@ export class Subscriber extends Component {
      * @param type Type. May be wildcarded or namespaced like in EventEmitter2.
      * @param listener Callback. Returns nothing.
      */
-    on<T extends Event>(
+    on<T extends Event = GenericEvent>(
```
Comment (andy-ms) [inline types/cote/index.d.ts]: I think this makes it too easy to forget a type argument. If you really want this to be unchecked you can write `send<any>`.

**Pull request closed without merging**